src/app/repositories/firebase/dataset_firebase_repository.py

```diff
@@ -35,44 +37,89 @@ def get_latest_dataset_with_survey_id_and_period_id(
 
         return dataset_metadata
 
-    def perform_new_dataset_transaction(
+    def perform_batched_dataset_write(
         self,
         dataset_id: str,
         dataset_metadata_without_id: DatasetMetadataWithoutId,
         unit_data_collection_with_metadata: list[UnitDataset],
         extracted_unit_data_identifiers: list[str],
-    ):
+    ) -> bool:
         """
-        Writes dataset metadata and unit data to firestore as a transaction, which is
-        rolled back if any of the operations fail.
-
+        Write dataset metadata and unit data to firestore in batches.
         Parameters:
-        dataset_id: id of the dataset
-        dataset_metadata_without_id: dataset metadata without a dataset id
-        unit_data_collection_with_metadata: collection of unit data associated to a dataset
-        extracted_unit_data_identifiers: identifiers associated with the unit data collection
+        dataset_id (str): The unique id of the dataset
+        dataset_metadata_without_id (DatasetMetadataWithoutId): The metadata of the dataset without its id
+        unit_data_collection_with_metadata (list[UnitDataset]): The collection of unit data associated with the new dataset
+        extracted_unit_data_identifiers (list[str]): List of identifiers ordered to match the identifier for each set of
+        unit data in the collection.
+
         """
 
-        # A stipulation of the @firestore.transactional decorator is the first parameter HAS
-        # to be 'transaction', but since we're using classes the first parameter is always
-        # 'self'. Encapsulating the transaction within this function circumvents the issue.
-        @firestore.transactional
-        def dataset_transaction(transaction: firestore.Transaction):
-            new_dataset_document = self.datasets_collection.document(dataset_id)
+        logger.info("Performing batch writes")
+        new_dataset_document = self.datasets_collection.document(dataset_id)
+        unit_data_collection_snapshot = new_dataset_document.collection("units")
+
+        try:
+            batch = self.client.batch()
+            batch.set(new_dataset_document, dataset_metadata_without_id, merge=True)
+            batch.commit()
+
+            batch_counter = 0
+            batch = self.client.batch()
+
+            for i in range(len(unit_data_collection_with_metadata)):
+                if batch_counter > 0 and batch_counter % self.BATCH_SIZE == 0:
+                    batch.commit()
+                    batch = self.client.batch()
+
+                new_unit = unit_data_collection_snapshot.document(
+                    extracted_unit_data_identifiers[i]
+                )
+                batch.set(new_unit, unit_data_collection_with_metadata[i], merge=True)
+                batch_counter += 1
+
+            batch.commit()
```

Comment:
this final commit will only be required if there is a remainder on the unit_data_collection_with_metadata % 500 where it will loop and never become divisible by 500 without a remainder. is there any effect to batch.commit() when the batch is empty? I wonder if there is any point is defining the batch size by survey type or size of data because these smaller surveys are probably not going to fill a batch

Reply:
This commit is always going to be needed as the other commit, the one that run between batches, will only serve when the data is > batch_size. Either we have data count that is just divisible by 500, or we have remainder data count for the last batch, this commit will serve its purpose to commit the last batch.
We are now taking the maximum size a batch can hold, which I think is the best because atomicity is preserved within a batch and we need not to run through the clean up process if anything exception occur within a single batch. That'll save time as the deletion is running quite slow.

Reply:
is it worth doing if not batch_size % 500 then batch.commit commit if there was a remainder then you will need to commit after the loop? or do you not see any point?

Reply:
Because the commit that run between batches will only be run when there is a number of record of batch_size % 500 + 1. Eg. if we have 501 records, then 1-500 batch will be committed and open another batch for the last record.
If we have exactly 500 records, this commit is still necessary to commit the batch.